### DISCUSSION TITLE
Inline event question

### DISCUSSION THREAD

**Discussion by YTN0** [mikke89/RmlUi#343] [Q&A]
Am new to RmlUI and am currently in the process of understanding how events work.

If I have an inline event e.g. <button onclick="blah">Start Game</button>, how do I get the value "blah" from the event is generated?

I have my custom event listener instancer which directs to my custom EventListener. I know I have the "blah" value in the instancer, but this doesn't seem to propagate through to the actual event that triggers in EventListener (unless I am missing something).

Thanks.

**Comment by mikke89** [comment]
Right, so, since you have the value in the instancer it is a matter of passing that along to the event listener. So for example, you might want to have one special event listener that handles "blah" events, e.g. `EventListenerBlah`. Then you instance a new `EventListenerBlah`, or pass on a pointer from a singleton if there only needs to be a single instance.

If you have another inline event you want to handle, let's say "blurgh", you can instance another type `EventListenerBlurgh` that handles such events.

If you just want to have a single event listener that handles everything based on the same string that is passed to the instancer, then you can make something like an `EventListenerGeneric` that takes a string on construction. Then it is a matter of instancing a new one of that type and pass the string into it.

**Reply by YTN0** [reply]
I was hoping to be able to have a single event listener for all events and then do the filtering of the parameter (e.g. blah) within that. 

It's not a huge deal... I have some workarounds I could use, but that would have been the easiest / cleanest approach.

Appreciate the thoughts on it.

**Reply by mikke89** [reply]
That's exactly what you can do with the `EventListenerGeneric` approach I described above. I mean you have multiple instances but only a single type and process function. You can also take a look at the Invader sample, this is the same approach used there.

**Reply by YTN0** [reply]
The generic option is a good one, but it's still multiple instances (albeit specific to the types e.g. blah)...

The reason for my preferring a single instance has to do with trying to create a reusable C# wrapper that I can then use for various cases.

A lot of my wrapper is based on having a singleton / static instance, and having a single instance of the event listener would have tied neatly into that.

**Reply by mikke89** [reply]
Right, I see. Although I guess then it's really only be a matter of passing the string + event on to the singleton / a free function. Then you have exactly what the Invader sample does, see the [EventManager](https://github.com/mikke89/RmlUi/blob/master/Samples/invaders/src/EventManager.cpp) and how it gets the arguments from the listener and instancer.

**Comment by YTN0** [comment]
Thanks for all the input. 

FYI, very nice job on the overall design of the system... It's very well thought out and cleanly put together, which makes it very easy to integrate with. 

I like how you abstracted out the various parts like system, render and file handling. It's clearly shows that a lot of thought went into it!

**Reply by mikke89** [reply]
No problem, happy to help out and discuss with new users.

These decisions go all the way back to the libRocket days, before my time, so all credit to the original developers. I agree these interfaces make it very nice to integrate into an existing project, which certainly is one of the things I originally liked about the library :)
